Add support for composite datasets

consoull opened the pull request:
### Linked issue(s):
Towards KOL-3558

### What change does this PR introduce and why?
Adds support for composite datasets with `DATAPOINT/COMPOSITE` value and tags nested datasets with their corresponding dataset type.

The below screenshot illustrates this change. The dataset consists of nested text datasets `a` and `b`. The top level `data_type` is marked `DATASET/COMPOSITE` while the nested datasets `a` and `b` have their `data_type` marked `DATASET/TEXT`
![Screenshot 2023-12-11 at 1 19 19 PM](https://github.com/kolenaIO/kolena/assets/25123885/f21c320e-1154-493c-819d-dcdef50ed0fb)

### Please check if the PR fulfills these requirements

- [x] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [x] Relevant tests for the changes have been added
- [x] Relevant docs have been added / updated

consoull committed:
Add support for composite datasets

Adds support for composite datasets with `DATAPOINT/COMPOSITE` value
and adds tag nested datasets with their corresponding dataset type.

consoull commented on kolena/_experimental/dataset/_dataset.py:
```diff
@@ -83,10 +84,21 @@ def _infer_datatype_value(x: str) -> str:
-    if FIELD_LOCATOR in df.columns:
-        return df[FIELD_LOCATOR].apply(_infer_datatype_value)
-    elif FIELD_TEXT in df.columns:
+def add_datatype(df: pd.DataFrame) -> None:
```
Marked public since this function is utilized by backend. This function opts to modify the passed DataFrame so the prefix parsing logic can be shared

ly-kolena replied:
I kinda disagree with this. This additional info is for internal use only, we don't expose this to users or want them to think about it. When users get retrieve the data back, they also don't see this temporary internal column.

consoull committed:
Add unit tests for `add_datatype`

yifanwu-kolena commented on tests/unit/_experimental/dataset/test_dataset.py:
```diff
@@ -52,6 +53,43 @@ def test__infer_datatype_value(uri: str, expected: str) -> None:
+    )
+
+
+def test__add_datatype__composite() -> None:
```
in my understanding, composite supports arbitrary amount of sub datapoints. should we test for 1 sub dp and more than 2 dps as well?

consoull replied:
Ya, your understanding is correct. I will add a test for 1 sub datapoint. Do you see benefit in having a test for 3 datapoints over 2?

consoull replied:
I'm going to add some edge cases which are missing from this test (eg. doubly nested composites)

yifanwu-kolena replied:
yeah, agree that 3 or 2 sub dp have no difference

consoull committed:
Update unit test to cover single composite datapoint

consoull committed:
Update composite dataset storage format

consoull commented on kolena/_experimental/dataset/_dataset.py:
```diff
@@ -83,6 +86,40 @@ def _infer_datatype_value(x: str) -> str:
+        df[DATA_TYPE_FIELD] = _infer_datatype(df)
+
+
+def _format_composite(df: pd.DataFrame, prefixes: Set[str], sep: str) -> None:
```
Not really a fan of the arguments here, but `_add_datatype` method was getting bulky. I'll likely update this to work on a single composite (prefix at a time and keep the for loop in `_add_datatype`

consoull committed:
Update `_format_composite` args

consoull committed:
Python 3.7 friendly

consoull committed:
Python 3.8 friendly

consoull committed:
Make seperator a constant

consoull committed:
Remove serialization of composites to JSON

consoull committed:
Check if `data_type` is member of DatapointType

Updates the check for `data_type` field to ensure the value is also in
the DatapointType enum. This is done to ensure the annotation
types (eg. `'data_type': 'ANNOTATION/LABEL'`) do not pass the check.

consoull committed:
Use `SEP` over literal '.'

ly-kolena commented on tests/unit/_experimental/dataset/test_dataset.py:
```diff
@@ -52,6 +57,89 @@ def test__infer_datatype_value(uri: str, expected: str) -> None:
+    with pytest.raises(InputValidationError):
+        _add_datatype(composite_dataset)
+
+
```
I added this test

```
def test__datapoint_dataframe__serde_composite() -> None:
    datapoints = [
        {
            "category": "outdoor" if i < 5 else "indoor",
            "is_same": True if i % 3 else False,
            "a.locator": f"https://test-iamge-{i}.png",
            "a.width": 500 + i,
            "a.height": 400 + i,
            "a.bboxes": [
                LabeledBoundingBox(label="car", top_left=[i, i], bottom_right=[i + 50, i + 50])
                for i in range(random.randint(2, 6))
            ],
            "a.label": ScoredClassificationLabel(label="dog", score=0.1 + i * 0.05),
            "b.locator": f"https://test-iamge-{i}.png",
            "b.width": 500 + i,
            "b.height": 400 + i,
            "b.bboxes": [
                LabeledBoundingBox(label="car", top_left=[i, i], bottom_right=[i + 50, i + 50])
                for i in range(random.randint(2, 6))
            ],
            "b.label": ScoredClassificationLabel(label="cat", score=0.1 + i * 0.05),
        }
        for i in range(10)
    ]
    df = pd.DataFrame(datapoints)
    df_expected = pd.DataFrame(
        dict(
            datapoint=[
                dict(
                    category=dp["category"],
                    is_same=dp["is_same"],
                    a={
                        DATA_TYPE_FIELD: DatapointType.IMAGE,
                        "label": dp["a.label"]._to_dict(),
                        "width": dp["a.width"],
                        "height": dp["a.height"],
                        "locator": dp["a.locator"],
                        "bboxes": [bbox._to_dict() for bbox in dp["a.bboxes"]],
                    },
                    b={
                        DATA_TYPE_FIELD: DatapointType.IMAGE,
                        "label": dp["b.label"]._to_dict(),
                        "width": dp["b.width"],
                        "height": dp["b.height"],
                        "locator": dp["b.locator"],
                        "bboxes": [bbox._to_dict() for bbox in dp["b.bboxes"]],
                    },
                    data_type=DatapointType.COMPOSITE,
                )
                for dp in datapoints
            ],
        ),
    )
    df_serialized = _to_serialized_dataframe(df, column=COL_DATAPOINT)

    assert df_serialized[COL_DATAPOINT].apply(json.loads).equals(df_expected[COL_DATAPOINT])

    df_expected = pd.DataFrame(
        [
            {
                "category": dp["category"],
                "is_same": dp["is_same"],
                "a.locator": dp["a.locator"],
                "a.width": dp["a.width"],
                "a.height": dp["a.height"],
                "a.bboxes": [
                    BoundingBox(label=bbox.label, top_left=bbox.top_left, bottom_right=bbox.bottom_right)
                    for bbox in dp["a.bboxes"]],
                "a.label": Label(label=dp["a.label"].label, score=dp["a.label"].score),
                "b.locator": dp["b.locator"],
                "b.width": dp["b.width"],
                "b.height": dp["b.height"],
                "b.bboxes": [
                    BoundingBox(label=bbox.label, top_left=bbox.top_left, bottom_right=bbox.bottom_right)
                    for bbox in dp["b.bboxes"]],
                "b.label": Label(label=dp["b.label"].label, score=dp["b.label"].score),
            }
            for dp in datapoints
        ]
    )
    df_deserialized = _to_deserialized_dataframe(df_serialized, column=COL_DATAPOINT)
    assert_frame_equal(df_deserialized, df_expected)
```

the final deserialization looks to have flattened `a.label` and `b.label`, which should remain `ClassificationLabel`.

consoull replied:
Thanks! This test uncovered a few issues. The recursive flattening is fixed by adding `max_level=0` option to https://pandas.pydata.org/docs/reference/api/pandas.json_normalize.html.

Another issue this uncovered is LabeledBoundingBox being cast to BoundingBox during deserialization. It appears many of the [subtypes (eg. LabeledBoundingBox, ScoredBoundingBox) are missing annotation types](https://github.com/kolenaIO/kolena/blob/trunk/kolena/workflow/annotation.py#L106) and being cast to their parent class due to inheriting the annotation label.
```
(Pdb) df_deser['a.bboxes'][0]
[BoundingBox(top_left=(0.0, 0.0), bottom_right=(50.0, 50.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0), BoundingBox(top_left=(1.0, 1.0), bottom_right=(51.0, 51.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0), BoundingBox(top_left=(2.0, 2.0), bottom_right=(52.0, 52.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0)]
(Pdb) df_orig['a.bboxes'][0]
[LabeledBoundingBox(top_left=(0.0, 0.0), bottom_right=(50.0, 50.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car'), LabeledBoundingBox(top_left=(1.0, 1.0), bottom_right=(51.0, 51.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car'), LabeledBoundingBox(top_left=(2.0, 2.0), bottom_right=(52.0, 52.0), width=50.0, height=50.0, area=2500.0, aspect_ratio=1.0, label='car')]
```

ly-kolena replied:
For the 2nd point, it is currently a known limitation for the tradeoff for ease of use, as we don't have enough information to deserialize to subtypes.

ly-kolena commented on kolena/_experimental/dataset/_dataset.py:
```diff
@@ -83,6 +90,35 @@ def _infer_datatype_value(x: str) -> str:
+            composite[DATA_TYPE_FIELD] = _infer_datatype(composite)
+
+            df[prefix] = composite.to_dict("records")
+            df.drop(columns=composite_columns, inplace=True)
```
Can we avoid using `inplace=True`. My general reading is that there is no clear cut performance benefit and with our usage of passing dataframe as input, it's easy to introduce hard-to-debug bugs because of inplace manipulation. And there's no way to tell function caller that we do that to their input.

Edit: Looks like we're manipulating df in general in this function anyway, gonna think if there's alternative, otherwise the above comment doesn't matter much/

ly-kolena replied:
Although weary of the usage of `inplace=True` and prefer not to have function that manipulates input dataframe, this comment is nonblocking, as currently `_add_datatype` has single invocation point and it is modifying a dataframe internally created, not user's dataframe.

consoull replied:
`_add_datatype` is going to have a few more invocations on the backend so I think it is worthwhile to have the discussion now.  We can update the function to return a new copy of the DataFrame leaving the previous one intact, but if all functions along the dataset deser call stack ensure the input dataframe is not manipulated there will certainly be performance implications for large datasets.

In my opinion, the public api `register_dataset` should create a copy immediately and called private functions are free to mutate. This ensures that the users DataFrame is not mutated and keeps copies to a minimum.

ly-kolena replied:
Even for private functions, they could be reused/shared by other public/private methods, now or future. A function that modifies input data is easy to cause bugs when call paths are non-trivial. Current scoping to `_add_datatype` (as small as possible) looks ok to me, still future developers should be aware of this behavior if they want to use it. I don't think we should expand it to `register_dataset`.

A slightly different discussion of use of `inplace=True`, I don't want to blindly believe in performance gain from `inplace`, but also don't want to unnecessarily fear of it. For dicussion's sake, references that in general not in favor of `inplace`
1. https://stackoverflow.com/questions/45570984/in-pandas-is-inplace-true-considered-harmful-or-not
2. https://sourcery.ai/blog/pandas-inplace/

ly-kolena commented on kolena/_experimental/dataset/_dataset.py:
```diff
@@ -95,7 +131,7 @@ def _infer_datatype(df: pd.DataFrame) -> Union[pd.DataFrame, str]:
     result = _dataframe_object_serde(df, _serialize_dataobject)
     if column == COL_DATAPOINT:
-        result[DATA_TYPE_FIELD] = _infer_datatype(df)
+        _add_datatype(result)
```
Current logic has no issue, but I'm wondering if we should add a test to ensure `_to_serialized_dataframe` doesn't apply composite heuristic to, say, serializing result dataframe.

ly-kolena commented on tests/unit/_experimental/dataset/test_dataset.py:
```diff
@@ -52,6 +57,89 @@ def test__infer_datatype_value(uri: str, expected: str) -> None:
+    with pytest.raises(InputValidationError):
+        _add_datatype(composite_dataset)
+
+    composite_dataset = pd.DataFrame(
```
Really really nit: a test case like this has unexpected output

```
    composite_dataset = pd.DataFrame(
        {
            ".dots": [i for i in range(5)],
            "x": [i for i in range(5)],
        },
    )
    _add_datatype(composite_dataset)
```

consoull replied:
I think this is a good callout! I will add a validation check for empty prefix

ly-kolena reviewed (approved):
With fix of max_level to `flatten_composite`, logic looks good.

consoull committed:
Set `pd.json_normalize(max_level=0)` to avoid recursive flattening

Fixes a bug where `pd.json_normalize` was recursively flattening nested
composite attributes. The desired level of flattening can be achieved
by setting the optional `max_level` arg to 0.
Adds a unit test which covers this case.

consoull committed:
Add input validation check for empty/whitespace prefix

consoull committed:
Add unit test to ensure composite serialization not applied to result
